pallets/rc-migrator/src/lib.rs

```diff
@@ -670,19 +718,100 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Update the total number of XCM messages processed by the Asset Hub.
+		/// Receive a query response from the Asset Hub for a previously sent xcm message.
 		#[pallet::call_index(3)]
-		#[pallet::weight(T::RcWeightInfo::update_ah_msg_processed_count())]
-		pub fn update_ah_msg_processed_count(origin: OriginFor<T>, count: u32) -> DispatchResult {
+		#[pallet::weight(T::RcWeightInfo::receive_query_response())]
+		pub fn receive_query_response(
+			origin: OriginFor<T>,
+			query_id: QueryId,
+			response: Response,
+		) -> DispatchResult {
+			match <T as Config>::ManagerOrigin::ensure_origin(origin.clone()) {
+				Ok(_) => {
+					// Origin is valid [`Config::ManagerOrigin`].
+				},
+				Err(_) => {
+					match <T as Config>::RuntimeOrigin::from(origin.clone()).into() {
+						Ok(pallet_xcm::Origin::Response(response_origin))
+							if response_origin == Location::new(0, Parachain(1000)) =>
+						{
+							// Origin is valid - this is a response from Asset Hub
+						},
+						_ => {
+							return Err(BadOrigin.into());
+						},
+					}
+				},
+			}
+
+			let response = match response {
+				Response::DispatchResult(maybe_error) => maybe_error,
+				_ => return Err(Error::<T>::InvalidQueryResponse.into()),
+			};
+
+			if matches!(response, MaybeErrorCode::Success) {
+				log::info!(
+					target: LOG_TARGET,
+					"Received success response for query id: {}",
+					query_id
+				);
+				PendingXcmMessages::<T>::remove(query_id);
+			} else {
+				log::error!(
+					target: LOG_TARGET,
+					"Received error response for query id: {}; response: {:?}",
+					query_id,
+					response.clone(),
+				);
+			}
+
+			Self::deposit_event(Event::<T>::QueryResponseReceived { query_id, response });
+
+			Ok(())
+		}
+
+		/// Resend a previously sent and unconfirmed XCM message.
+		#[pallet::call_index(4)]
+		#[pallet::weight(T::RcWeightInfo::resend_xcm())]
+		pub fn resend_xcm(origin: OriginFor<T>, query_id: u64) -> DispatchResultWithPostInfo {
 			<T as Config>::ManagerOrigin::ensure_origin(origin)?;
-			Self::update_msg_processed_count(count);
+
+			let xcm = PendingXcmMessages::<T>::get(query_id).ok_or(Error::<T>::QueryNotFound)?;
+
+			if let Err(err) = send_xcm::<T::SendXcm>(Location::new(0, Parachain(1000)), xcm) {
+				log::error!(target: LOG_TARGET, "Error while sending XCM message: {:?}", err);
+				Self::deposit_event(Event::<T>::XcmResendAttempt {
+					query_id,
+					send_error: Some(err),
+				});
+			} else {
+				Self::deposit_event(Event::<T>::XcmResendAttempt { query_id, send_error: None });
+			}
+
+			Ok(Pays::No.into())
```

Comment:
Why was `Ok(Pays::No.into())` added only here and not to the other calls?
All of those calls are guarded by `ManagerOrigin::ensure_origin`, and I expect them to be called with `Unpaid + Transact` from other chains, right?

But if we plan to allow `ManagerOrigin` for specific accounts so they can call these extrinsics directly (without XCM from other chain), then we should add `Ok(Pays::No.into())` to all of them. Otherwise, when we migrate balances, those accounts won't be able to call.

Reply:
good point. I will do it. but also those accounts still would need some balance. I actually would filter that account and not migrate it on RC